### PR TITLE
Track incremental parse depth

### PR DESCRIPTION
**Pull request by mattwar**
Fixes issue #13398 

Add ability to track the number of incremental parses queued up on a document and switch to using full parsing if the queue gets too long.

@heejaechang @Pilchie @jasonmalinowski @CyrusNajmabadi  please review

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs [hunk 1]
```diff
@@ -172,14 +186,23 @@ private static string GetSyntaxTreeFilePath(DocumentInfo info)
             return TreeAndVersion.Create(tree, textAndVersion.Version);
         }
 
+        private static int GetIncrementalParseDepth(ValueSource<TreeAndVersion> treeSource)
+        {
+            var incrementalTreeSource = treeSource as IncrementalTreeSource;
+            return incrementalTreeSource != null ? incrementalTreeSource.GetIncrementalParseDepth() : 0;
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs]: `return incrementalTreeSource?.GetIncrementalParseDepth() ?? 0;`

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs [hunk 2]
```diff
@@ -387,6 +410,9 @@ internal DocumentBranch(SourceText text, DocumentState state)
             }
         }
 
+        // This number is arbitrary. It is intended to keep the chain of incremental parses relatively small
+        private const int MAX_INCREMENTAL_PARSE_DEPTH = 32;
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs]: I'm quite leery of checking this in without have decent perf testing of it in advance.

**Comment by Pilchie** [comment]
The implementation looks correct to me, but I worry about the performance consequences of this change.  @rchande @KevinH-MS are there any good ways to see what sort of impact this might have on typing performance?

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs]: I actually prefer Matt's version. I just can't get used to these 'as' expression idioms.

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs]: I would like this to be an option, and not hardcoded in. That way we can adjust things without delivering a new binary if there's a problem.

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState_IncrementalTreeSource.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                _current = current;
+                _previous = previous;
+            }
+
+            /// <summary>
+            /// Computes the number of incremental parses that will occur in order to realize the current source's tree.
+            /// </summary>
+            public int GetIncrementalParseDepth()
+            {
+                int depth = 0;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState_IncrementalTreeSource.cs]: Could depth just be computed up front and stored in the instance?  Seems like that would be simpler?

**Comment by CyrusNajmabadi** [comment]
Is there a reason long incremental queue depths are very undesirable?  Seems like in such a scenario, doing a full parse might be bad as it could cause a lot of allocations.

**Comment by CyrusNajmabadi** [comment]
I'm trying to understand this comment:

>My problem is... In XS if we do many(talking about 1000+(formatting full file or something like that on huge file)) changes to file... Roslyn than incrementally parses all of this changes which results in stack overflow since it's requesting ParseIncrementally from AsyncTask which is chaining until stackoverflow occurs...

1. Why is formatting causing 1000 parses?
2. Can we avoid overflow by looping instead of recursing?

**Comment by CyrusNajmabadi** [comment]
Also, what is XS

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs (hunk 2 above)
Comment (mattwar) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs]: Me either.. Just wanted some feedback first.

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState_IncrementalTreeSource.cs (hunk 3 above)
Comment (mattwar) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState_IncrementalTreeSource.cs]: I tried that first, but that ends up just causing a forced full parse each N text updates regardless of the typing behavior.  I'm trying to short circuit that if incremental parses are actually happening often enough.

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState_IncrementalTreeSource.cs]: Good point.  Please document as such.  That's a subtlety that could easily be lost.

**Comment by heejaechang** [comment]
should we add test or something to prevent this logic to be deleted again?

**Comment by mattwar** [comment]
@CyrusNajmabadi long parse chains occur if you rapidly accrue lots of text updates between parses.  This could be caused by a refactoring that uses text changes a lot, or really fast typing before IDE features catch up.

Obviously really long chains then lead to stack overflow when then parsing does catch up.  However, its also bad for performance when the number of incremental parses in the chain are more expensive to perform than a full parse.  However, I don't yet know what the number is.

**Comment by heejaechang** [comment]
it would be nice if incremental parse request in not opaque like current one but info on incremental parsing is actually exposed. then next incremental parse request could aggregate span to incremental parse rather than keep chaining. and remove chain length tracking and let it automatically turn to full parse when many place of file has changed.

for typing case, most likely part changed that will be aggregated will be close enough so perf probably won't be affected. for refactoring case like this, perf won;t better that much anyway.

**Comment by DavidKarlas** [comment]
@CyrusNajmabadi Reason XS causes 1000 changes to file on formatting is that we request formatting and then iterate over all changes in formatted file and apply those to file 1 by 1.. We do that because in past Roslyn wasn't good at handling non-Windows new line(\n) and this was workaround... Maybe we can remove this... I'm not 100% sure...

**Comment by CyrusNajmabadi** [comment]
> Reason XS causes 1000 changes to file on formatting is that we request formatting and then iterate over all changes in formatted file and apply those to file 1 by 1.. 

Yeah... that's not good :)

It would be much better to make a new SourceText with WithChanges passing all those changes in.

> We do that because in past Roslyn wasn't good at handling non-Windows new line(\n) and this was workaround... 

I'd love to know what sort of issue you were running into.  Seems like we should fix that instead of trying to work around it.  As Roslyn wants to be great for Linux and Mac we def want to handle \n just as good as we handle \r\n.

Thanks!

**Comment by CyrusNajmabadi** [comment]
I'd prefer we not do this work, and just have XS not submit thousands of changes for something like formatting.  @DavidKarlas how are you getting those formatting edits in the first place?  Are you getting them through our Formatter API?

**Comment by DavidKarlas** [comment]
I think \n and \r\n problem was solved... We will probably be able to formatting issue on XS side... Here is example how we do this atm.. https://github.com/mono/monodevelop/blob/master/main/src/addins/CSharpBinding/MonoDevelop.CSharp.Formatting/OnTheFlyFormatter.cs#L106-L114

**Comment by CyrusNajmabadi** [comment]
Note: instead of calling: ```foreach (var change in newTree.GetChanges (syntaxTree)) {``` you can just call FormatAsync and then call GetFormattedTextChanges on the result.  This will give you the list of changes for you to apply yourself.  Not sure if that makes your life any easier.

**Comment by mattwar** [comment]
I would still prefer to have this change and avoid possible stack overflows in the future..  You know, really fast typing.

**Comment by jaredpar** [comment]
We apologize, but we are closing this PR due to code drift. We regret letting this PR go so long without attention, but at this point the code base has changed too much for us to revisit older PRs. Going forward, we will manage PRs better under an SLA currently under draft in issue #26266 – we encourage you to add comments to that draft as you see fit. Although that SLA is still in draft form, we nevertheless want to move forward with the identification of older PRs to give contributors as much early notice as possible to review and update them.

If you are interested in pursuing this PR, please reset it against the head of master and we will reopen it. Thank you!

**Pull request closed without merging**